api/v1alpha1/hostedcluster_types.go

```diff
@@ -35,6 +35,9 @@ type HostedClusterSpec struct {
 
 	// InfraID is used to identify the cluster in cloud platforms
 	InfraID string `json:"infraID,omitempty"`
+
+	ControlPlaneServiceType                string `json:"controlPlaneServiceType,omitempty"`
+	ControlPlaneServiceTypeNodePortAddress string `json:"controlPlaneServiceTypeNodePortAddress,omitempty"`
```

Comment:
First, thanks for the contributions! As the surface area continues to expand I want to suggest we start being more deliberate about our changes to the API. In thinking through this, I bet it's likely if we accept this concept that we'll end up wanting more abstractly defined behaviors and configurations for those behaviors going forward. We've been through something very similar in OCP with the [ingress publishing API](https://github.com/openshift/api/blob/master/operator/v1/types_ingress.go#L417), and I'd like to see if those ideas could help us given it underwent significant up front design and revision.

So at a minimum, if we can agree on the idea that a control plane is "published" using some strategy, I would ask we consider a discriminated union type like `EndpointPublishingStrategy` which has clearly defined selectable behaviors and also provides the means for us to configure them.

So, maybe something like:

```go
type HostedCluster struct {
	// ...
	PublishingStrategy ControlPlanePublishingStrategy
}

type ControlPlanePublishingStrategy struct {
	Type ControlPlanePublishingStrategyType `json:"type"`

	LoadBalancer *LoadBalancerStrategy `json:"loadBalancer,omitempty"`

	NodePort *NodePortStrategy `json:"nodePort,omitempty"`
}
```

If someone else has a different way of characterizing the actual behaviors here besides a "publishing strategy" or another idea entirely about how we should go about it (e.g. a way to avoid needing it at all, or finding some unified way so we don't need multiple ways, etc.) I'd love to hear more.

cc @sjenning @enxebre @csrwng @derekwaynecarr @zanetworker

Reply:
I am open for whatever way y'all think is best: I will wait to do another refactor until we get an idea on it however.

Reply:
I do like the idea of a `PublishingStrategy`, where the `NodePortStrategy` contains the external address to associate with nodeports. 
@relyt0925 right now, besides the API server and VPN, we rely on routes on the management cluster to expose oauth and machine config. Is that something that would work for a ROKS cluster?

Reply:
@csrwng it will work from a technical perspective. The issue here is a few things from an IBM perspective:

We are looking for a path similar to what we have in place today where we do not need to provision load balancer services on a per cluster basis (it does not work in our scale model). We are ultimately looking to be able to have a constant amount of "load balancers" (3 in the typical case one per zone) multiplex all nodeport traffic over it. 

There's more advanced use cases we can discuss with you going forward as well but that is the next step for us to start moving forward in adoption. 

I'm basically looking to add a path to where per cluster ingress controllers and load balancers are not required.

Reply:
This doesn't fully get us there: but it's a step to basically adding a path to allow that.

I view that there will be multiple strategies over time. Currently the default now is load balancer services for apiserver and vpn and then per cluster ingress controllers that serve specific routes for the management cluster.

Another one that would require work in the future but might be applicable is picture a world when konnectivity is deployed (everything is TCP at that point) and we use SNI and all traffic can flow through a set router/istio gateway through separate unique domains that can be used to point to the appropriate backend. I'm sure there are others but those are the two I know off the top of my head (the benefit of that approach is there would no longer be a need to expose the entire node port range in the management cluster (could just be 443 and then per cluster config is added to a proxy that can handle routing to specific backends).

Reply:
> Currently the default now is load balancer services for apiserver and vpn and then per cluster ingress controllers that serve specific routes for the management cluster.

Just to clarify... the per-cluster ingress controllers is currently just a hack to make ingress work for guest clusters. That will go away when we move to ingress on the workers. The routes we configure for oauth and ignition work through the single router of the management cluster, so they don't really incur a per-cluster cost.

Reply:
You are correct then (that would not be applicable). It would just be the loadBalancer services.